### PR TITLE
tracing: fix opentelemetry service name

### PR DESCRIPTION
**Pull request by zirain**
fix: https://github.com/istio/istio/issues/42080

- fix missing opentelemetry service name
- fix e2e test

opentelemetry send zipkin span as following:
```json
[
{
      "traceId": "b72c4f57c2cf8dc73df8ea734dcc1481",
      "parentId": "6f5703b3bf77bc1d",
      "id": "266fd6ae4b9bb617",
      "kind": "SERVER",
      "name": "ingress",
      "timestamp": 1669102679177576,
      "duration": 1399,
      "localEndpoint": {
        "serviceName": "httpbin.default"
      },
      "tags": {
        "component": "proxy",
        "downstream_cluster": "-",
        "guid:x-request-id": "233128f2-8a20-929e-bf22-843422da5bde",
        "http.method": "GET",
        "http.protocol": "HTTP/1.1",
        "http.status_code": "200",
        "http.url": "http://httpbin.default:8000/ip",
        "istio.canonical_revision": "v1",
        "istio.canonical_service": "httpbin",
        "istio.mesh_id": "cluster.local",
        "istio.namespace": "default",
        "node_id": "sidecar~10.244.0.46~httpbin-847f64cc8d-982r2.default~default.svc.cluster.local",
        "peer.address": "10.244.0.47",
        "provider": "otel",
        "request_size": "0",
        "response_flags": "-",
        "response_size": "28",
        "upstream_cluster": "inbound|80||",
        "upstream_cluster.name": "inbound|80||",
        "user_agent": "curl/7.86.0-DEV",
        "zone": ""
      }
    },
    {
      "traceId": "b72c4f57c2cf8dc73df8ea734dcc1481",
      "id": "6f5703b3bf77bc1d",
      "kind": "CLIENT",
      "name": "egress httpbin.default:8000",
      "timestamp": 1669102679177281,
      "duration": 1962,
      "localEndpoint": {
        "serviceName": "sleep.default"
      },
      "tags": {
        "component": "proxy",
        "downstream_cluster": "-",
        "guid:x-request-id": "233128f2-8a20-929e-bf22-843422da5bde",
        "http.method": "GET",
        "http.protocol": "HTTP/1.1",
        "http.status_code": "200",
        "http.url": "http://httpbin.default:8000/ip",
        "istio.canonical_revision": "latest",
        "istio.canonical_service": "sleep",
        "istio.mesh_id": "cluster.local",
        "istio.namespace": "default",
        "node_id": "sidecar~10.244.0.47~sleep-69cfb4968f-6qw9g.default~default.svc.cluster.local",
        "peer.address": "10.244.0.47",
        "provider": "otel",
        "request_size": "0",
        "response_flags": "-",
        "response_size": "28",
        "upstream_cluster": "outbound|8000||httpbin.default.svc.cluster.local",
        "upstream_cluster.name": "outbound|8000||httpbin.default.svc.cluster.local",
        "user_agent": "curl/7.86.0-DEV",
        "zone": ""
      }
    }
]
```

### CODE REVIEW

**Comment by hanxiaop** [comment]
Verified the fix works.

File: tests/integration/telemetry/tracing/otelcollector/testdata/otel-tracing.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  name: otel-tracing
+spec:
+  tracing:
+  - providers:
+    - name: test-otel
+    randomSamplingPercentage: 100.0
+    customTags:
+      "provider":
+        literal:
+          value: "otel"
```
Comment (aimuz) [inline tests/integration/telemetry/tracing/otelcollector/testdata/otel-tracing.yaml:13]: Add a blank line?

**Comment by hanxiaop** [comment]
This actually makes different `OpenTelemetyConfig` configurations of different addresses within one proxy config dump all have the same service name , which is it's node cluster name.

File: pilot/pkg/networking/core/v1alpha3/tracing.go [hunk 2]
```diff
@@ -369,6 +369,7 @@ func otelConfigGen(_, hostname, cluster string) (*anypb.Any, error) {
 				},
 			},
 		},
+		ServiceName: serviceName,
```
Comment (hanxiaop) [inline pilot/pkg/networking/core/v1alpha3/tracing.go:372]: This actually makes different OpenTelemetyConfig configurations of different addresses within one proxy config dump all have the same service name , which is it's node cluster name. Is this the correct behavior?

Comment (zirain) [reply pilot/pkg/networking/core/v1alpha3/tracing.go:372]: this will represent in `localEndpoint.ServiceName`

Comment (zirain) [reply pilot/pkg/networking/core/v1alpha3/tracing.go:372]: opencensus:
```
2022-11-22T11:36:50.206Z	DEBUG	loggingexporter/logging_exporter.go:52	ResourceSpans #0
Resource SchemaURL: 
Resource labels:
     -> service.name: STRING(sleep.default)
     -> opencensus.starttime: STRING(2022-11-22T11:36:50.203009581Z)
     -> host.name: STRING(sleep-69cfb4968f-drx45)
     -> process.pid: INT(26)
     -> telemetry.sdk.version: STRING(0.5.0-dev)
     -> opencensus.exporterversion: STRING(0.5.0-dev)
     -> telemetry.sdk.language: STRING(cpp)
```

opentelemetry:
```
Resource SchemaURL: 
Resource labels:
     -> service.name: STRING(httpbin.default)
ScopeSpans #0
ScopeSpans SchemaURL: 
InstrumentationScope  
Span #0
    Trace ID       : 5cf4b6b70a6243c504b4dd7cda1c2a60
    Parent ID      : 9bc416e8a6b058ec
    ID             : 08d21cd30af47aa0
    Name           : ingress
    Kind           : SPAN_KIND_SERVER
    Start time     : 2022-11-22 11:35:59.208882 +0000 UTC
    End time       : 2022-11-22 11:35:59.210737 +0000 UTC
    Status code    : STATUS_CODE_UNSET
    Status message : 
Attributes:
     -> node_id: STRING(sidecar~10.244.0.109~httpbin-847f64cc8d-5pnw5.default~default.svc.cluster.local)
     -> zone: STRING()
     -> guid:x-request-id: STRING(125c09f3-8959-9948-b9b7-46ab2ac4b2e0)
     -> http.url: STRING(http://httpbin:8000/ip)
     -> http.method: STRING(GET)
     -> downstream_cluster: STRING(-)
     -> user_agent: STRING(curl/7.86.0-DEV)
     -> http.protocol: STRING(HTTP/1.1)
     -> peer.address: STRING(10.244.0.110)
     -> request_size: STRING(0)
     -> response_size: STRING(28)
     -> component: STRING(proxy)
     -> upstream_cluster: STRING(inbound|80||)
     -> upstream_cluster.name: STRING(inbound|80||)
     -> http.status_code: STRING(200)
     -> response_flags: STRING(-)
     -> istio.mesh_id: STRING(cluster.local)
     -> istio.canonical_service: STRING(httpbin)
     -> istio.namespace: STRING(default)
     -> istio.canonical_revision: STRING(v1)
     -> provider: STRING(otel)
```

**Review by hanxiaop — APPROVED** [review]

**Review by lei-tang — APPROVED** [review]

**Review by hzxuzhonghu — APPROVED** [review]

**Review by kfaseela — APPROVED** [review]

**Comment by radekg** [comment]
This doesn't exist in 1.16.0, correct? Hence my pods crashing like this

```
2022-12-02T16:53:15.326028Z     warning envoy config    gRPC config for type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error adding/updating listener(s) 10.246.155.212_4194: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'
0.0.0.0_8080: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'
10.246.157.211_8888: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'
10.246.157.238_8083: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'
0.0.0.0_3000: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'
```

is expected. Is this correct?

**Comment by zirain** [comment]
> This doesn't exist in `1.16.0`, correct? Hence my pods crashing with:
> 
> ```
> 2022-12-02T16:53:15.326028Z     warning envoy config    gRPC config for type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error adding/updating listener(s) 10.246.155.212_4194: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'
> 0.0.0.0_8080: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'
> 10.246.157.211_8888: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'
> 10.246.157.238_8083: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'
> 0.0.0.0_3000: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'
> ```
> 
> is expected. Correct?

yes, will land in `1.16.1`

**Comment by zakariais** [comment]
I am using 1.16.1 but still seeing this issue.
Any updates on what could be wrong?
`"scope":"ads","msg":"ADS:LDS: ACK ERROR docker-http-https-echo-5c9c7678b5-kkr2c.devops-apps-2164 Internal:Error adding/updating listener(s) 10.191.183.62_3001: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n172.20.189.9_4318: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.178.228_3001: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.203.43_4194: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.181.20_3001: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.217.68_4194: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.182.78_4194: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.191.60_8042:`

**Comment by zirain** [comment]
> I am using 1.16.1 but still seeing this issue. Any updates on what could be wrong? `"scope":"ads","msg":"ADS:LDS: ACK ERROR docker-http-https-echo-5c9c7678b5-kkr2c.devops-apps-2164 Internal:Error adding/updating listener(s) 10.191.183.62_3001: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n172.20.189.9_4318: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.178.228_3001: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.203.43_4194: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.181.20_3001: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.217.68_4194: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.182.78_4194: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.191.60_8042:`

did you restart the deployment?

**Comment by zakariais** [comment]
@zirain yes i did restart but still seeing above issue in 1.16.1

**Comment by zakariais** [comment]
I updated the version to 1.16.4 but still i am getting above/below issue
`{"level":"warn","time":"2023-04-07T12:49:05.323371Z","scope":"ads","msg":"ADS:LDS: ACK ERROR docker-http-https-echo-5c9c7678b5-kkr2c.devops-apps-22 Internal:Error adding/updating listener(s) 10.191.193.146_3001: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.198.242_4194: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n172.20.164.8_8888: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.187.254_3001: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n0.0.0.0_9153: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.215.142_4194: Didn't find a registered implementation for name: 'envoy.tracers.opentelemetry'\n10.191.217.148_4194: Didn't find a registered implementation for name:`
The config i am using is
```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istio-control-plane
  namespace: istio-system
spec:
  namespace: istio-system
  profile: default
  meshConfig:
    defaultProviders:
      tracing:
      - "otel"
    enableTracing: true
    defaultConfig:
      tracing:
        sampling: 100
    extensionProviders:
    - name: "otel"
      opentelemetry:
        service: "otel-collector.monitoring.svc.cluster.local"
        port: 4317
```

**Comment by zirain** [comment]
can you run following command:

```bash
kubectl get po <your_pod_name>. -oyaml | grep image
```

**Comment by zakariais** [comment]
@zirain 
```
image: docker.io/istio/pilot:1.16.4
imagePullPolicy: IfNotPresent
image: docker.io/istio/pilot:1.16.4
imageID: docker.io/istio/pilot@sha256:6b239dc64dc767242f7871f195dcd96b832a5e7b7db945c4505782ca04dcb71a
```

**Comment by zirain** [comment]
what about sidecar? did you restart dataplane?

**Comment by zakariais** [comment]
@zirain 
image: 'docker.io/istio/proxyv2:1.16.4'
We checked the sidecar/dataplane as well, the versions are correct.
I did lots of restarts as well.
Still i am getting the same warn

**Comment by zirain** [comment]
can you run `istioctl bug-report` and share me the output(via slack if you want)?

**Pull request merged**